File: frameworks/dotnet/aspnet-mvc/Implementation/Controllers/ContractController.cs
using System.Reflection;
using System.Runtime.InteropServices;
using Microsoft.AspNetCore.Mvc;

namespace Implementation.Controllers;

/// <summary>/health and /__meta, which the contract asks of every target outside the corpus.</summary>
[ApiController]
public sealed class ContractController : ControllerBase
{
    // The payloads are loaded before the server starts, so a server that answers has them.
    [HttpGet("/health")]
    public ContentResult Health() => Content("ok", "text/plain");

    [HttpGet("/__meta")]
    public Meta About() => new("ASP.NET Core MVC", Version(typeof(ControllerBase).Assembly), RuntimeInformation.FrameworkDescription, "", Boot.Ms);

    /// <summary>
    /// The version of the MVC assembly the shared framework loaded, without the source revision
    /// after the plus.
    /// </summary>
    private static string Version(Assembly assembly)
    {
        string? informational = assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
        int plus = informational?.IndexOf('+') ?? -1;
        return plus > 0 ? informational![..plus] : informational ?? assembly.GetName().Version?.ToString() ?? "";
    }
}

File: frameworks/node/nestjs/Implementation/contract/contract.controller.ts
import { readFileSync } from 'node:fs';

import { Controller, Get, Header, Inject } from '@nestjs/common';

import { PLATFORM, type Platform } from '../platform.js';

/** The version of the Nest core that was installed, read from its package, which exports no package.json. */
const version = (JSON.parse(readFileSync(new URL('package.json', import.meta.resolve('@nestjs/core')), 'utf8')) as { version: string }).version;

/** /health and /__meta, which the contract asks of every framework outside the corpus. */
@Controller()
export class ContractController {
  constructor(@Inject(PLATFORM) private readonly platform: Platform) {}

  // The payloads are loaded before Nest creates the application, so an application that answers has them.
  @Get('health')
  @Header('Content-Type', 'text/plain; charset=utf-8')
  health(): string {
    return 'ok';
  }

  @Get('__meta')
  meta() {
    return {
      framework: 'NestJS',
      version,
      runtime: `Node.js ${process.versions.node}`,
      adapter: this.platform.adapter,
      serializer: 'JSON.stringify',
    };
  }
}
